dags/google_api_helper.py

```diff
@@ -124,11 +126,51 @@ def ramp_down_cluster(key, total_size):
         target_sizes = Variable.get("cluster_target_size", deserialize_json=True)
         target_sizes[key] = total_size
         Variable.set("cluster_target_size", target_sizes, serialize_json=True)
-        reduce_instance_group_size(key, total_size)
+
+        min_size = get_min_size(key)[0]
+        if total_size >= min_size:
+            reduce_instance_group_size(key, total_size)
     except:
         reduce_instance_group_size(key, total_size)
 
 
+def get_min_size(
+    key: Optional[str] = None, min_sizes: Optional[dict[str, tuple[int, str]]] = None
+) -> int:
+    """Determines the minimum size to enforce on a cluster.
+
+    A minimum size value is considered active if its timestamp hasn't passed.
+    The min_sizes argument comes from the cluster_min_size Variable. The user
+    can pass this dictionary to reduce the number of database queries.
+    """
+    if min_sizes is None:
+        try:
+            min_sizes = Variable.get("cluster_min_size", deserialize_json=True)
+        except:
+            return 0, min_sizes
+
+    if key is None or key not in min_sizes:
+        return 0, min_sizes
+
+    # a standard time to use when the timestamp has expired and
+    # we shouldn't notify the user (we already have, or the user never asked)
+    y2k = datetime(2000, 1, 1)
+
+    min_size, rawtimestamp = min_sizes[key]
+    timestamp = datetime.fromisoformat(rawtimestamp)
+
+    if datetime.utcnow() > timestamp:
+
+        if timestamp != y2k:  # expired, but we haven't notified the user yet
+            slack_message(f":ice_cube: warm up period for {key} has expired")
+            min_sizes[key] = [0, y2k.isoformat()]
+            Variable.set("cluster_min_size", min_sizes, serialize_json=True)
+
+        return 0, min_sizes
+
+    return min_size, min_sizes
```

Comment:
Is the second item returned used anywhere, it seems every call only want the first one.

Reply:
The original idea was if this function modifies `min_sizes`, then you might want an updated copy, but you're probably right for now. I can just remove it and it's easy to add back if it ever becomes useful.

Reply:
OK, I think remove the extra return is a good idea. Actually since the `min_sizes` is passed into the function as a dictionary, Should any change to the dict remained after the call? I remember python do pass by reference for all the lists and dicts. so the min_sizes is updated anyway. That might be an unwanted behavior though.

Reply:
That's true. It is passed by reference. Another reason to remove it.